### PR TITLE
Patch 1

### PR DESCRIPTION
**Pull request by such00bd**
Review

### CODE REVIEW

File: IssueService.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+public class IssueService {
+
+	@Autowired
+	private IssueRepository issueRepository;
+
+	public Issue update(Issue issue) throws InvalidIssueStatusException, IssueUpdateException{
+
+		// TODO: Validate the issue.status first;
+		// Status can be updated from CREATED to IN_PROGRESS to COMPLETED.
+		if(issue ==  null || issueRepository == null ||
```
Comment (vinaylakkam) [inline IssueService.java]: `issueRepository` will never be null as it's injected by Spring; If it is null then application won't get bootstrapped at all; pls remove this.

File: IssueService.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+	@Autowired
+	private IssueRepository issueRepository;
+
+	public Issue update(Issue issue) throws InvalidIssueStatusException, IssueUpdateException{
+
+		// TODO: Validate the issue.status first;
+		// Status can be updated from CREATED to IN_PROGRESS to COMPLETED.
+		if(issue ==  null || issueRepository == null || 
+				(getIssueStatus(issue.getStatus()) - getIssueStatus(issueRepository.getCurrentStatus(issue)) != 1)){
```
Comment (vinaylakkam) [inline IssueService.java]: Move this line to a method (say `isValidStatus`) to make it clean.

File: IssueService.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+			throw new InvalidIssueStatusException("Issue status is invalid!");
+		}		
+
+		// TODO: Save the issue using issueRepository
+		issueRepository.updateIssue(issue);
+	}
+	
+	private int getIssueStatus(String issueRepoStatus){
+		if("CREATED".equalsIgnoreCase(issueRepoStatus)){
```
Comment (vinaylakkam) [inline IssueService.java]: Use enums for status

File: IssueService.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+		if(issue ==  null || issueRepository == null || 
+				(getIssueStatus(issue.getStatus()) - getIssueStatus(issueRepository.getCurrentStatus(issue)) != 1)){
+			throw new InvalidIssueStatusException("Issue status is invalid!");
+		}		
+
+		// TODO: Save the issue using issueRepository
+		issueRepository.updateIssue(issue);
+	}
+	
+	private int getIssueStatus(String issueRepoStatus){
```
Comment (vinaylakkam) [inline IssueService.java]: i think variable name can just be `status` or `issueStatus`

File: IssueService.java (hunk 2 above)
Comment (vinaylakkam) [inline IssueService.java]: Also status may or may not change; The user may just update issue description, but don't change the status.

File: IssueService.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+		// TODO: Validate the issue.status first;
+		// Status can be updated from CREATED to IN_PROGRESS to COMPLETED.
+		if(issue ==  null || issueRepository == null || 
+				(getIssueStatus(issue.getStatus()) - getIssueStatus(issueRepository.getCurrentStatus(issue)) != 1)){
+			throw new InvalidIssueStatusException("Issue status is invalid!");
+		}		
+
+		// TODO: Save the issue using issueRepository
+		issueRepository.updateIssue(issue);
```
Comment (vinaylakkam) [inline IssueService.java]: you want to `return` the issue right

File: IssueService.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+@Service
+public class IssueService {
+
+	@Autowired
+	private IssueRepository issueRepository;
+
+	public Issue update(Issue issue) throws InvalidIssueStatusException, IssueUpdateException{
```
Comment (vinaylakkam) [inline IssueService.java]: `IssueUpdateException` ?

**Comment by such00bd** [comment]
updated

File: IssueService.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	 * 
+	 * @param issue
+	 * @return
+	 */
+	private boolean isValidStatus(Issue issue) {
+		try {
+			if (issue == null
+					|| issue.getStatus() == null
+					|| (IssueStatus.valueOf(issue.getStatus().toUpperCase())
+							- IssueStatus.valueOf(issueRepository.getCurrentStatus(issue).toUpperCase()) < 0)) {
```
Comment (vinaylakkam) [inline IssueService.java]: I don't think issueRepository will have getCurrentStatus(issue) method. Instead it generally has get(issueId) where you can get current issue and retrieve the status.

File: IssueService.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	
+	/**
+	 * Checks if issue status is valid to update
+	 * 
+	 * @param issue
+	 * @return
+	 */
+	private boolean isValidStatus(Issue issue) {
+		try {
+			if (issue == null
```
Comment (vinaylakkam) [inline IssueService.java:30]: if issue is null, we are throwing `InvalidIssueStatusException` from `update` method? I think this null condition should go into `update` method only

Comment (such00bd) [reply IssueService.java:30]: first step is validation and then update, if issue is null why we will call update?

File: IssueService.java (hunk 7 above)
Comment (such00bd) [reply IssueService.java]: written based on assumption that IssueRepository will take Issue object as input and provide the status, we can update this based on IssueRepository functionality.

File: IssueServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+	@InjectMocks
+	IssueService issueService = new IssueService();
+
+	@Mock
+	IssueRepository issueRepository;
+
+
+	@Test
+	public void testSuccessUpdate() {
```
Comment (vinaylakkam) [inline IssueServiceTest.java:20]: Please see better naming standards: https://dzone.com/articles/7-popular-unit-test-naming

File: IssueServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		Issue repoIssue = new Issue();
+		repoIssue.setIssueId(123);
+		repoIssue.setStatus("CREATED");
+		Issue issueUpdate = new Issue();
+		issueUpdate.setIssueId(123);
+		issueUpdate.setStatus("IN_PROGRESS");
+		Mockito.when(issueRepository.getIssue(issueUpdate)).thenReturn(repoIssue);
```
Comment (vinaylakkam) [inline IssueServiceTest.java:27]: getIssue accepts issueId no?

File: IssueServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+		issueService.update(issueUpdate);
+		Mockito.verify(issueRepository).getIssue(issueUpdate);
+		Mockito.verify(issueRepository, never()).update(issueUpdate);
+	}
+	
+	@Test(expected = InvalidIssueStatusException.class)
+	public void testUpdateFail() {
+		Issue repoIssue = new Issue();
+		repoIssue.setIssueId(123);
+		repoIssue.setStatus("CREATED");
```
Comment (vinaylakkam) [inline IssueServiceTest.java:53]: Above 3 lines are duplicate code in other methods

File: IssueService.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+					|| issue.getStatus() == null
+					|| (IssueStatus.valueOf(issue.getStatus().toUpperCase())
+							- IssueStatus.valueOf(getRepoIssueStatus(issue).toUpperCase()) < 0)) {
+				return false;
+
+			}
+		} catch (IllegalArgumentException excep) {
+			return false;
```
Comment (vinaylakkam) [inline IssueService.java:38]: Is this code covered in tests?

**Review by vinaylakkam — COMMENTED** [review]
Your tests are great. Just try to improve readability next time. They say that 'Test cases serve the purpose of documentation.. meaning if you read the test cases you should understand what the component (being tested) does.'

**Pull request merged**